kyo-core/shared/src/main/scala/kyo/typeMap.scala

```diff
@@ -5,48 +5,51 @@ import scala.collection.immutable.HashMap
 
 opaque type TypeMap[+A] = HashMap[Tag[Any], Any]
 
-extension [A](self: TypeMap[A])
+object TypeMap:
+    extension [A](self: TypeMap[A])
 
-    private inline def fatal[T](using t: Tag[T]): Nothing =
-        throw new RuntimeException(s"fatal: kyo.TypeMap of contents [${self.show}] missing value of type: [${t.showTpe}].")
+        private inline def fatal[T](using t: Tag[T]): Nothing =
+            throw new RuntimeException(s"fatal: kyo.TypeMap of contents [${self.show}] missing value of type: [${t.showTpe}].")
 
-    def get[B >: A](using t: Tag[B]): B =
-        val b = self.getOrElse(t.erased, null)
-        if !isNull(b) then b.asInstanceOf[B]
-        else
-            var sub: B = null.asInstanceOf[B]
-            val it     = self.iterator
-            try // iterator should never throw given type constraint on B
-                while isNull(sub) do
-                    val (tag, item) = it.next()
-                    if tag <:< t then
-                        sub = item.asInstanceOf[B]
-                end while
-            catch
-                case _: NoSuchElementException => fatal
-            end try
-            sub
-        end if
-    end get
+        def get[B >: A](using t: Tag[B]): B =
+            val b = self.getOrElse(t.erased, null)
+            if !isNull(b) then b.asInstanceOf[B]
+            else
+                var sub: B = null.asInstanceOf[B]
+                val it     = self.iterator
+                try // iterator should never throw given type constraint on B
+                    while isNull(sub) do
+                        val (tag, item) = it.next()
+                        if tag <:< t then
+                            sub = item.asInstanceOf[B]
+                    end while
+                catch
+                    case _: NoSuchElementException => fatal
+                end try
+                sub
+            end if
+        end get
 
-    inline def add[B](b: B)(using inline t: Tag[B]): TypeMap[A & B] =
-        self.updated(t.erased, b)
+        inline def add[B](b: B)(using inline t: Tag[B]): TypeMap[A & B] =
+            self.updated(t.erased, b)
 
-    inline def union[B](that: TypeMap[B]): TypeMap[A & B] =
-        self ++ that
+        inline def union[B](that: TypeMap[B]): TypeMap[A & B] =
+            self ++ that
 
-    inline def size: Int        = self.size
-    inline def isEmpty: Boolean = self.isEmpty
+        def prune[B >: A](using t: Tag[B]): TypeMap[B] =
+            if t =:= Tag[Any] then empty
```

Comment:
Technically yes - and I am happy to make that change for 'soundness'.

I intentionally took this departure from `ZEnvironment` as I think it's likely more useful.

What can you `.get` from a `TypeMap[Any]`? I suppose you could invoke `.get[Any]` to get the first element? But that does not seem useful nor practical. The widening of the `A` -> `Any` renders the resulting map practically useless.

Am I missing a usecase?

Reply:
If you check the tests, you can see that
1. Intersection eventually should be supported
2. Pruning to single type may not result in a single value.

Reply:
I'm not sure if I understood the semantics well but would `filter` be a better name for this new method? Regarding `Any`, maybe we could fail compilation in case it's passed via `scala.compiletime.error`?

Reply:
I was attempting to keep similar naming to ZEnv, though that's not necesssary. I think `prune` is likely the most accurate. I think `.filter` would to support a predicate.